CRTX-81731: Support rendering SLA field in timer cell

lmichalevich opened the pull request:
#### Description

We want SLA fields to look as close as they can to the web-client.

![Screenshot 2023-05-29 at 19 32 02](https://github.com/demisto/sane-reports/assets/73780437/073403b3-c009-434e-afc8-b5ed2e329b67)

---------

#### Generated JSON
```
[
  {
    "type": "table",
    "data": [
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "1",
        "Name": "Test",
        "Remediation SLA": "Status: idle"
      },
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "3",
        "Name": "test 3",
        "Remediation SLA": "Status: running ()"
      },
      {
        "Containment SLA": "Status: ended ()",
        "Detection SLA": "Status: running ()",
        "ID": "2",
        "Name": "test",
        "Remediation SLA": "Status: idle"
      }
    ],
    "layout": {
      "classes": "striped stackable small compact",
      "columnPos": 0,
      "forceRangeMessage": "",
      "h": 3,
      "i": "dd2917b0-fd67-11ed-9bb3-6bd79a4877db",
      "readableHeaders": {
        "Containment SLA": "Containment SLA",
        "Detection SLA": "Detection SLA",
        "Remediation SLA": "Remediation SLA",
        "id": "ID",
        "name": "Name"
      },
      "reflectDimensions": true,
      "rowPos": 0,
      "tableColumns": [
        "id",
        "name",
        "Detection SLA",
        "Containment SLA",
        "Remediation SLA"
      ],
      "tableColumnsMetaData": [
        {
          "displayed": true,
          "isDefault": true,
          "key": "id",
          "width": 110
        },
        {
          "displayed": true,
          "isDefault": true,
          "key": "name",
          "width": 100
        },
        {
          "displayed": true,
          "key": "Detection SLA",
          "width": 200
        },
        {
          "displayed": true,
          "key": "Containment SLA",
          "width": 163
        },
        {
          "displayed": true,
          "key": "Remediation SLA",
          "width": 200
        }
      ],
      "w": 12
    },
    "query": {
      "type": "incident",
      "filter": {
        "query": "",
        "period": {
          "by": "",
          "byTo": "",
          "byFrom": "days",
          "toValue": null,
          "fromValue": 7,
          "field": ""
        },
        "fromDate": "0001-01-01T00:00:00Z",
        "toDate": "0001-01-01T00:00:00Z"
      }
    },
    "automation": {
      "name": "",
      "id": "",
      "args": null,
      "noEvent": false
    },
    "fromDate": "2023-05-22T16:39:57Z",
    "title": "Test",
    "emptyNotification": "No results found",
    "titleStyle": null,
    "autoPageBreak": true
  }
]
```

#### Generated Report - Before

[Report.pdf](https://github.com/demisto/sane-reports/files/11593597/incidentDailyReportTempalte.pdf)

#### Generated Report - After

[Report.pdf](https://github.com/demisto/sane-reports/files/11593604/incidentDailyReportTempalte.pdf)

------------- 

#### In this PR

Adding new field to the generated JSON named `extraData`.

#### BE changes

In order to see these changes we need to add the `extraData` section to the generated json.
it is an array of objects. each object contains the relevant information to render the time cell (see the example in the description). We also need to add inside each item the type of the cell. (type: 'timer').

The information comes from the following apis:

![Screenshot 2023-05-30 at 13 11 23](https://github.com/demisto/sane-reports/assets/73780437/9f3cdd54-1ae6-4749-b251-10a6f82bd8a8)

threshold:

![Screenshot 2023-05-30 at 13 15 49](https://github.com/demisto/sane-reports/assets/73780437/219893cc-0162-4a56-8464-9aefe4c30e24)


#### Example JSON
```
[
  {
    "type": "table",
    "data": [
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "1",
        "Name": "Test",
        "Remediation SLA": "Status: idle"
      },
      {
        "Containment SLA": "Status: idle",
        "Detection SLA": "Status: running ()",
        "ID": "3",
        "Name": "test 3",
        "Remediation SLA": "Status: running ()"
      },
      {
        "Containment SLA": "Status: ended ()",
        "Detection SLA": "Status: running ()",
        "ID": "2",
        "Name": "test",
        "Remediation SLA": "Status: idle"
      }
    ],
    "layout": {
      "classes": "striped stackable small compact",
      "columnPos": 0,
      "forceRangeMessage": "",
      "h": 3,
      "i": "dd2917b0-fd67-11ed-9bb3-6bd79a4877db",
      "readableHeaders": {
        "Containment SLA": "Containment SLA",
        "Detection SLA": "Detection SLA",
        "Remediation SLA": "Remediation SLA",
        "id": "ID",
        "name": "Name"
      },
      "reflectDimensions": true,
      "rowPos": 0,
      "tableColumns": [
        "id",
        "name",
        "Detection SLA",
        "Containment SLA",
        "Remediation SLA"
      ],
      "tableColumnsMetaData": [
        {
          "displayed": true,
          "isDefault": true,
          "key": "id",
          "width": 110
        },
        {
          "displayed": true,
          "isDefault": true,
          "key": "name",
          "width": 100
        },
        {
          "displayed": true,
          "key": "Detection SLA",
          "width": 200
        },
        {
          "displayed": true,
          "key": "Containment SLA",
          "width": 163
        },
        {
          "displayed": true,
          "key": "Remediation SLA",
          "width": 200
        }
      ],
      "w": 12
    },
    "query": {
      "type": "incident",
      "filter": {
        "query": "",
        "period": {
          "by": "",
          "byTo": "",
          "byFrom": "days",
          "toValue": null,
          "fromValue": 7,
          "field": ""
        },
        "fromDate": "0001-01-01T00:00:00Z",
        "toDate": "0001-01-01T00:00:00Z"
      }
    },
    "extraData": [
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:34:26.447968+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 20,
          "slaStatus": -1,
          "startDate": "2023-05-28T18:14:26.447968+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 30,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 0
        }
      },
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:18:40.697968+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 2,
          "slaStatus": -1,
          "startDate": "2023-05-28T18:16:40.697968+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 30,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-06-03T19:28:32.280921+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "2023-05-29T19:28:32.280921+03:00",
          "totalDuration": 0,
          "threshold": 0
        }
      },
      {
        "Detection SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-28T18:15:28.339413+03:00",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "running",
          "sla": 20,
          "slaStatus": -1,
          "startDate": "2023-05-28T17:55:28.339413+03:00",
          "totalDuration": 0,
          "threshold": 0
        },
        "Containment SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "2023-05-29T19:56:29.235357+03:00",
          "endDate": "2023-05-29T19:29:31.44581+03:00",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "ended",
          "sla": 30,
          "slaStatus": 1,
          "startDate": "2023-05-29T19:26:29.235357+03:00",
          "totalDuration": 182,
          "threshold": 3
        },
        "Remediation SLA": {
          "type": "timer",
          "accumulatedPause": 0,
          "breachTriggered": false,
          "dueDate": "0001-01-01T00:00:00Z",
          "endDate": "0001-01-01T00:00:00Z",
          "lastPauseDate": "0001-01-01T00:00:00Z",
          "runStatus": "idle",
          "sla": 7200,
          "slaStatus": -1,
          "startDate": "0001-01-01T00:00:00Z",
          "totalDuration": 0,
          "threshold": 0
        }
      }
    ],
    "automation": {
      "name": "",
      "id": "",
      "args": null,
      "noEvent": false
    },
    "fromDate": "2023-05-22T16:39:57Z",
    "title": "Test",
    "emptyNotification": "No results found",
    "titleStyle": null,
    "autoPageBreak": true
  }
]
```

#### Generated Report - Before

[Report.pdf](https://github.com/demisto/sane-reports/files/11593590/incidentDailyReportTempalte.pdf)

#### Generated Report - After

[Report.pdf](https://github.com/demisto/sane-reports/files/11593663/incidentDailyReportTempalte.pdf)

yanyo87 commented:
![image](https://github.com/demisto/sane-reports/assets/61918543/d16ff8f4-d446-444e-b7e0-e7f78965d19b)
remove these files

lmichalevich commented:
Tested on another JSON:

[test.json](https://github.com/demisto/sane-reports/files/11599488/test.txt)

[after.pdf](https://github.com/demisto/sane-reports/files/11599466/after.pdf)
[before.pdf](https://github.com/demisto/sane-reports/files/11599470/before.pdf)

yanyo87 commented on src/css/index.less:
```diff
@@ -0,0 +1,13 @@
+* {
+  letter-spacing: 0.1px;
+  -webkit-font-smoothing: antialiased;
+}
+
+i {
+  &.demisto-icon {
+    font-size: 24px !important;
+    display: inline-block;
+    margin: 0 !important;
+    vertical-align: sub !important;
+  }
+}
```
don't do this globally,
add the relevant part only to the timer cell

lmichalevich replied:
moved them to the timer cell

yanyo87 commented on src/utils/time.js:
```diff
@@ -31,3 +31,95 @@ export function isDateNotValid(d) {
+}
+
+export function getSlaProps(timerValue) {
+  const shouldTick = timerValue.runStatus === TIMER_STATUS.running;
+  const timePeriod = EXTENDED_TIME_PERIODS.SECONDS;
+  const duration = getSlaDurationOrNull(timerValue);
+  const dueDate = timerValue.dueDate;
+  const startDate = getSlaEffectiveStartTime(timerValue);
+  const riskThreshold = timerValue.threshold || 0;
+
+  return { shouldTick, timePeriod, duration, startDate, dueDate, riskThreshold };
```
```suggestion

  const duration = getSlaDurationOrNull(timerValue);
  const dueDate = timerValue.dueDate;
  const startDate = getSlaEffectiveStartTime(timerValue);
  const riskThreshold = timerValue.threshold || 0;

  return { shouldTick: timerValue.runStatus === TIMER_STATUS.running, timePeriod :EXTENDED_TIME_PERIODS.SECONDS , duration ...., startDate, dueDate, riskThreshold };
```

yanyo87 commented on src/utils/time.js:
```diff
@@ -31,3 +31,95 @@ export function isDateNotValid(d) {
+}
+
+function getSlaTimeLeft(timerValue) {
+  const effectiveStopDate = getSlaEffectiveStopDate(timerValue);
+
+  return moment(timerValue.dueDate).diff(effectiveStopDate, EXTENDED_TIME_PERIODS.SECONDS);
```
```suggestion
  return moment(timerValue.dueDate).diff(getSlaEffectiveStopDate(timerValue), EXTENDED_TIME_PERIODS.SECONDS);
```

lmichalevich replied:
It is like this in the web-client. I think both of the way to right code are ok (no preferred way).

lmichalevich commented:
Updated the reports.

Added the SLA time with ellipsis (the format is too long and otherwise it looks bad when it breaks)

#### Before

[incidentDailyReportTempalte.pdf](https://github.com/demisto/sane-reports/files/11674851/incidentDailyReportTempalte.pdf)

#### After

[incidentDailyReportTempalte.pdf](https://github.com/demisto/sane-reports/files/11674834/incidentDailyReportTempalte.pdf)
